fdbserver/StorageMetrics.actor.cpp

```diff
@@ -548,10 +548,10 @@ void StorageServerMetrics::add(KeyRangeMap<int>& map, KeyRangeRef const& keys, i
 }
 
 // Returns the sampled metric value (possibly 0, possibly increased by the sampling factor)
-int64_t TransientStorageMetricSample::addAndExpire(KeyRef key, int64_t metric, double expiration) {
-	int64_t x = add(key, metric);
+int64_t TransientStorageMetricSample::addAndExpire(const Key& key, int64_t metric, double expiration) {
+	auto x = add(key, metric);
 	if (x)
-		queue.emplace_back(expiration, std::make_pair(*sample.find(key), -x));
+		queue.emplace_back(expiration, std::make_pair(key, -x));
```

Comment:
I see that in one of the classes, the queue for expiration holds an iterator, but in this one it holds a key, so another find() will have to be done during expiration.  Can that also be changed to an iterator?

Reply:
I don't think we can save the iterator here. Because the `add` function returns the delta not the real sum of metrics on that key. In side `add`, if the updated sum is 0, the sampler will clear that key so the iterator returned is `end()`, which makes it impossible to add the delta back to the original key after expiration.

Reply:
> this class is used for other sample types where the Metric can be increased or decreased for a key

It's possible in the future. But now I think in all of our cases `metrics` is a non-negative number. If we are good to add an `ASSERT_GE(metrics, 0)`. We can do this optimization.

Reply:
@sfc-gh-etschannen any thoughts?